### ISSUE TITLE
`Add Another Model` creates no nested inlines

### ISSUE THREAD

**Issue by koralarts** [datahub/grappelli-nested-inlines#10]
There seems to be a bug where the nested inlines inside an inline don't show up when you click `Add Another Model`. This seems to be affecting both the `NestedStackedInline` and `NestedTabularInline`.

This also affects the inlines when you set it to `extra = 0`.

**Comment by cwyalpha** [comment]
When click Add Another Model, the `addButtonHandler` in js add the hidden `grp-empty-form` in stacked.html (and tabular.html) to the end. The grappelli-nested-inlines defines the last formset of  `recursive_formset` as `grp-empty-form`, but the last formset doesn't have `nested_formsets`, so the line
 `{% if inline_admin_form.form.nested_formsets %}`
in stacked.html not works for the `grp-empty-form`.

I'm wondering is there a neat solution for this...

**Comment by koralarts** [comment]
Yeah I found that out while I was going through the calls. I looked for solutions inside the other repos but have not found any solutions. I will look into it a bit more because it makes the nested inlines kind of unusable with this happening.

**Comment by koralarts** [comment]
After a little more digging, I found that the nested inlines are not being added inside the inline because of a `for` loop that is checking the current forms inside the `formset`.

```
def add_nested_inline_formsets(self, request, inline, formset, depth=0):
        if depth > 5:
            raise Exception("Maximum nesting depth reached (5)")
        for form in formset.forms:
```

When `extra = 0` the `formset.forms` returns an empty list and doesn't go through the process of adding the nested inlines inside the inlines.

**Comment by cwyalpha** [comment]
Any solution for this?

**Comment by kosz85** [comment]
I think I found one, thought it's not full yet :) I used it before knowing it problems with adding new rows at inlined inlines.
Problem is not `add_nested_inline_formset` but it's place where it can be fixed :)
The main problem is that iterator of formset `formset.__iter__` is adding empty form with `__prefix__` that is replaced then by some nr. But this `formset.empty_form` is not having `nested_formsets`, and is not wrapped properly.

Here is monkeypatch, that you can add in your `models.py`, the best way is to add it to first model that initialize :) Before you use it, be sure to read it. It do some quite ogly things :D
- It forcefully overwrites property `empty_form` with prepared object
- It use what is in `wrap_nested_inline_formsets` and use it on empty_form to prepare it properly
- It doesn't solve double `__prefix__` problem in fields names - it's JS problem (only first should be replaced)
- I'm using it with other monkeypatch that solves this issue #9 

```
#Fix bug
import grappelli_nested.admin

def add_nested_inline_formsets(self, request, inline, formset, depth=0):
    if depth > 5:
        raise Exception("Maximum nesting depth reached (5)")
    empty_form = formset.empty_form
    for form in formset.forms + [empty_form]:
        nested_formsets = []
        for nested_inline in inline.get_inline_instances(request):
            InlineFormSet = nested_inline.get_formset(request, form.instance)
            prefix = "%s-%s" % (form.prefix, InlineFormSet.get_default_prefix())
            #because of form nesting with extra=0 it might happen, that the post data doesn't include values for the formset.
            #This would lead to a Exception, because the ManagementForm construction fails. So we check if there is data available, and otherwise create an empty form
            keys = request.POST.keys()
            has_params = any(s.startswith(prefix) for s in keys)
            if request.method == 'POST' and has_params:
                nested_formset = InlineFormSet(
                                request.POST,
                                request.FILES,
                                save_as_new="_saveasnew" in request.POST,
                                instance=form.instance,
                                prefix=prefix,
                                queryset=nested_inline.queryset(request)
                            )
            else:
                nested_formset = InlineFormSet(instance=form.instance,
                                               prefix=prefix, queryset=nested_inline.queryset(request))
            nested_formsets.append(nested_formset)
            if nested_inline.inlines:
                self.add_nested_inline_formsets(request, nested_inline, nested_formset, depth=depth+1)
        form.nested_formsets = nested_formsets
    form = empty_form
    media = None
    def get_media(extra_media):
        if media:
            return media + extra_media
        else:
            return extra_media
    wrapped_nested_formsets = []
    for nested_inline, nested_formset in zip(inline.get_inline_instances(request), form.nested_formsets):
         if form.instance.pk:
                instance = form.instance
         else:
            instance = None
         fieldsets = list(nested_inline.get_fieldsets(request))
         readonly = list(nested_inline.get_readonly_fields(request))
         prepopulated = dict(nested_inline.get_prepopulated_fields(request))
         wrapped_nested_formset = grappelli_nested.admin.InlineAdminFormSet(nested_inline, nested_formset,
             fieldsets, prepopulated, readonly, model_admin=self)
         wrapped_nested_formsets.append(wrapped_nested_formset)
         media = get_media(wrapped_nested_formset.media)
         if nested_inline.inlines:
             media = get_media(self.wrap_nested_inline_formsets(request, nested_inline, nested_formset))
    form.nested_formsets = wrapped_nested_formsets
    formset.__class__.empty_form = empty_form

grappelli_nested.admin.NestedModelAdmin.add_nested_inline_formsets = add_nested_inline_formsets
```

I'm not adding it as a fix, as it's ugly :) and it monkeypatch `empty_form` property. This won't be a problem in most cases, althought it's not clean solution.

**Comment by kosz85** [comment]
Modifing `grappelli_nested/static/grappelli_nested/js/grp_nested_inline.js` at line 109 from

```
            // update form index
            var re = /__prefix__/g;
```

to

```
            // update form index
            var re = /__prefix__/;
```

Do the job for JavaScript. So it replace only first `__prefix__` so inlines empty forms are usable. Thought it may be not backward compatible if you use custom templates with more complicated id structures. But for me it works.

**Comment by nareshdvd** [comment]
I am using solution provided by kosz85 for monkeypatching empty_form property, it surely adds nested inline form but the problem is when I click on "Add Another Model" for nested inline, it is always adding new inline form to empty form and not in the group where I clicked on  "Add Another Model". 
Any solution for this??

**Comment by kosz85** [comment]
Use that for Javascript also, as this one is responsible for adding new nested inline entry. Check if you have set proper JS. I had to add smth like this:

```
class OrderEntryInline(nested_admin.NestedTabularInline):
    template = 'admin/core/edit_inline/tabular.html'
    class Media:
        js = ('js/monkeypatch/grp_nested_inline.js',)
```

Where `monkeypatch` is fixed `grappelli_nested/static/grappelli_nested/js/grp_nested_inline.js` check also if you are using proper templates (I had to overwrite it to made it work properly - but no changes there). It's fully working for me now, so if you have any problems just ask. You may need to `collectstatic`.

**Comment by kosz85** [comment]
I found some other issues :) with 'Add another model' I used only one form, that was in extra with many inlined tabular forms which was just working, tought it doesn't work when you add new inlined form (with nested inlines). I found why, and I fixed it for my forms, thought it's fixed only for my custom form.
Oh, and one notice, every pregenerated inline form would just work, so it can be some kind of solution to have several `extra` inlined forms hidden.

Problems:
- `grp_nested_inlines.js:addButtonHandler()` is initiated for empty-forms which is wrong, I added one line solution to exit when it's empty-form: 

```
addButtonHandler = function(elem, options) {
        if (options.prefix.split('__prefix__').length != 1) return;
```
- `id` attr of form for `addButtonHandler` function is malformed (it doesn't have dash) fix:

```
            form.removeClass(options.emptyCssClass)
                .attr("id", empty_template.attr('id').replace("-empty", '-' + index));
```
- if you have autocomplete fields then they are not initialized, same with `grp_inline` I did recurention and it's working thought autocomplete fields are setuped per view and per form in custom template. It needs some kind of store for info about inline setup (fk_autocomplete, m2m_autocomplete, etc)

Maybe I should just do some commit to the lib.
